packages/lit-dev-cloudbuild-cleanup/src/cleanup.ts

```diff
@@ -0,0 +1,272 @@
+      console.log(`fetching ${serviceName} service`);
+      const {data} = await run.projects.locations.services.get({
+        name: `projects/${PROJECT_ID}/locations/${_DEPLOY_REGION}/services/${serviceName}`,
+      });
+
+      if (!data.spec || !data.spec.traffic) {
+        throw new Error(`found no traffic for ${serviceName} service`);
+      }
+
+      console.log(`${serviceName} traffic count`, data.spec.traffic.length);
+
+      const trafficToKeep = data.spec.traffic.filter((t) => {
+        return (
+          // precence of percent indicates traffic is being routed here (i.e. prod)
+          Boolean(t.percent) ||
+          revisionsToKeepByRecency.has(t.revisionName!) ||
+          openPrs.has(parseInt(t.tag!.slice(2), 10))
+        );
+      });
+
+      console.log(`${serviceName} traffic to keep count`, trafficToKeep.length);
+
+      if (trafficToKeep.length === 0) {
+        throw new Error(`found no ${serviceName} traffic to keep`);
+      }
+
+      if (data.spec.traffic.length !== trafficToKeep.length) {
+        console.log(`updating ${serviceName} service`);
+        data.spec.traffic = trafficToKeep;
+        await run.projects.locations.services.replaceService({
+          name: `projects/${PROJECT_ID}/locations/${_DEPLOY_REGION}/services/${serviceName}`,
+          requestBody: data,
+        });
+        serviceUpdated = true;
+      }
+    })
+  );
+
+  // revision status takes some time to update
+  if (serviceUpdated) {
+    console.log('waiting 30 seconds for traffic tags to update');
+    await sleep(30_000);
+
+    // fetch new list of revisions to get the now untagged (inactive) ones
+    console.log('fetching new revisions list');
+    const {data} = await run.projects.locations.revisions.list({
+      parent: `projects/${PROJECT_ID}/locations/${_DEPLOY_REGION}`,
+    });
+    revisions = data;
+  }
+
+  if (!revisions.items) {
+    throw new Error('found no new revision items');
+  }
+
+  console.log('looking for inactive revisions to delete');
+  const imageDigestsToKeep = new Set<string>();
+  for (const rev of revisions.items) {
+    if (!rev.metadata) {
+      throw new Error('revision has no metadata');
+    }
+
+    if (!rev.status || !rev.status.conditions) {
+      throw new Error('revision has no status conditions');
+    }
+
+    if (!rev.spec || !rev.spec.containers) {
+      throw new Error('revision has no containers');
+    }
+
+    if (
+      rev.status.conditions.find((cond) => cond.type === 'Active')!.status !==
+        'True' &&
+      // extra timestamp check in case the revision was just created and not active because of that
+      new Date(rev.metadata.creationTimestamp!) < ONE_WEEK_AGO
+    ) {
+      console.log(`deleting revision ${rev.metadata.name}`);
+      await run.projects.locations.revisions.delete({
+        name: `projects/${PROJECT_ID}/locations/${_DEPLOY_REGION}/revisions/${rev.metadata.name}`,
+      });
+      // wait 1 second to not hit api limit
```

Comment:
https://cloud.google.com/run/quotas#api
For writes, API limit is 60 per 60 seconds and I've hit that during testing without the wait. This was also more of an issue when we had hundreds of revisions to delete initially. I suppose we could also chunk the requests to groups of 60 and wait 1 minute in between too but figured this is easier. When run on a schedule, this likely won't have more than 60 to delete at a time so maybe it would be okay to not wait, but also there's no reason to really rush this either. WDYT?